### ISSUE TITLE
Meson falls back to bundled fluid synth because of unresolved readline dependency

### ISSUE THREAD

**Issue by PatTheMav** [dosbox-staging/dosbox-staging#1582]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

branch main

### What Operating System are you using?

Mac OS

### Relevant hardware info

x86_64, macOS Monterey

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Fluidsynth has a private dependency on `readline`, which is installed by Homebrew as "Keg only", meaning that it is downloaded and set up, but not added to default search paths of e.g. `pkg-config` or CMake:

> readline is keg-only, which means it was not symlinked into /usr/local, because macOS provides BSD libedit

Because of this, when relying on Homebrew to install build dependencies, meson will always try and _fail_ to find a pre-existing fluidsynth installation on macOS and use the fallback, which is _not_ automatically updated (as it's a onetime download).

Meson runs `pkg-config --cflags fluidsynth`, which fails with:
```
Package readline was not found in the pkg-config search path.
Perhaps you should add the directory containing `readline.pc'
to the PKG_CONFIG_PATH environment variable
Package 'readline', required by 'fluidsynth', not found
```

This can lead to more issues, as meson will then download the required dependencies, but because they are "static" downloads (and the _existence_ of the directories and their contained build scripts seems to suffice) they will - over time - not be in sync with assumptions made by dosbox-staging's meson build script, leading to errors like this:

```
subprojects/fluidsynth-2.2.4/meson.build:1:0: ERROR: In subproject fluidsynth: Unknown options: "fluidsynth:openmp"
```

In my case fluidsynth was downloaded a few months ago, but fetching and merging from upstream won't update the fluidsynth subdirectory of course.

*Note:* Running `PKG_CONFIG_PATH=/usr/local/opt/readline/lib/pkgconfig meson setup build` fixes this issue, as the pkconfig path for the "bottled" headline dependency is resolved by the environment variable specifying the additional search path.

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Be on macOS
2. Check out dosbox-staging
3. Install dependencies via Homebrew as outlined by `docs/build-macos.md`
4. Run `meson setup build`

### Provide a Log

```text
The Meson build system
Version: 0.61.2
Source dir: /Users/***/GitHub/dosbox-staging
Build dir: /Users/***/GitHub/dosbox-staging/build
Build type: native build
Project name: dosbox-staging
Project version: 0.79.0
C compiler for the host machine: ccache cc (clang 13.0.0 "Apple clang version 13.0.0 (clang-1300.0.29.30)")
C linker for the host machine: cc ld64 711
C++ compiler for the host machine: ccache c++ (clang 13.0.0 "Apple clang version 13.0.0 (clang-1300.0.29.30)")
C++ linker for the host machine: c++ ld64 711
Host machine cpu family: x86_64
Host machine cpu: x86_64
Has header "filesystem" : YES 
Compiler for C supports arguments -Wmaybe-uninitialized: NO 
Compiler for C++ supports arguments -Wmaybe-uninitialized: NO 
Compiler for C supports arguments -Weffc++: YES 
Compiler for C++ supports arguments -Weffc++: YES 
Compiler for C supports arguments -Wextra-semi: YES 
Compiler for C++ supports arguments -Wextra-semi: YES 
Compiler for C supports arguments -shared-libstdc++: NO 
Compiler for C++ supports arguments -shared-libstdc++: NO 
Compiler for C supports arguments -shared-libgcc: NO 
Compiler for C++ supports arguments -shared-libgcc: NO 
Compiler for C supports arguments -fPIC: YES 
Compiler for C++ supports arguments -fPIC: YES 
Checking for function "clock_gettime" : YES 
Checking for function "__builtin_available" : YES 
Checking for function "__builtin___clear_cache" : YES 
Checking for function "mprotect" : YES 
Checking for function "mmap" : YES 
Header <sys/mman.h> has symbol "MAP_JIT" : YES 
Checking for function "pthread_jit_write_protect_np" : YES 
Checking for function "sys_icache_invalidate" : YES 
Run-time dependency threads found: YES
Checking for function "pthread_setname_np" with dependency threads: YES 
Checking for function "realpath" : YES 
Checking whether type "struct dirent" has member "d_type" : YES 
Has header "pwd.h" : YES 
Has header "strings.h" : YES 
Has header "netinet/in.h" : YES 
Has header "sys/socket.h" : YES 
Checking for function "getpeername" : YES 
Checking for function "getsockname" : YES 
Checking if "test for setpriority support" compiles: YES 
Checking if "test for __builtin_expect support" compiles: YES 
Library dl found: YES
Library atomic found: NO
Library stdc++fs found: NO
Found pkg-config: /usr/local/bin/pkg-config (0.29.2)
Run-time dependency opusfile found: YES 0.12
Dependency threads found: YES unknown (cached)
Run-time dependency sdl2 found: YES 2.0.20
Run-time dependency sdl2_net found: YES 2.0.1
Run-time dependency gl found: YES
Found CMake: /usr/local/bin/cmake (3.22.2)
Run-time dependency fluidsynth found: NO (tried pkgconfig, framework and cmake)
Looking for a fallback subproject for the dependency fluidsynth
Building fallback subproject with default_library=shared

Executing subproject fluidsynth 

Message: You can disable this dependency with: -Duse_fluidsynth=false

subprojects/fluidsynth-2.2.4/meson.build:1:0: ERROR: In subproject fluidsynth: Unknown options: "fluidsynth:openmp"
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by shermp** [comment]
@kcgen Meson 0.60 made unknown options a [fatal error](https://mesonbuild.com/Release-notes-for-0-60-0.html#unknown-options-are-now-always-fatal). Since the `fluidsynth:openmp` option was only added in the `2.2.4-3` wrap, it should probably not be set in our `default_options` list.

@PatTheMav you can delete the `subprojects/fluidsynth-2.2.*` directory and `subprojects/packagecache/fluidsynth_2.2.*patch.zip` and `subprojects/packagecache/fluidsynth-2.2.*.zip` files to force meson to download an up-to-date version of the fluidsynth wrap.

Unfortunately, there isn't a lot we can do about homebrew shipping a broken fluidsynth library.

**Comment by PatTheMav** [comment]
@shermp Yes manually deleting the old folders and having meson download fresh copies fixes that particular issue.

As for the Homebrew issue, it might make sense to remove that requirement from the macOS docs then as the package seems only intended for use as a dylib.

**Comment by eli-schwartz** [comment]
> @PatTheMav you can delete the `subprojects/fluidsynth-2.2.*` directory and `subprojects/packagecache/fluidsynth_2.2.*patch.zip` and `subprojects/packagecache/fluidsynth-2.2.*.zip` files to force meson to download an up-to-date version of the fluidsynth wrap.

You can run `meson subprojects update --reset fluidsynth` and it will force re-extract the wrap.

You don't care if the packagecache files are there, they either have to be correct or they have the wrong filename because `subprojects.fluidsynth.wrap` was updated to point to new versions of the patch.zip -- either way, Meson will check them against the hash embedded in the .wrap file.

When `git pull` updates the .wrap file to point to a new upstream release of fluidsynth, the fluidsynth tarball will get updated to a new url, and the directory name of the extracted wrap will also change. At this point, Meson will no longer look for `subprojects/fluidsynth-2.2.4/` but e.g. `subprojects/fluidsynth-2.2.5/`, fail to find it, and have to extract it for the first time anyway.

When the WrapDB updates the patch file *only*, is where the issue comes in. The subproject directory is "correct" but the meson.build needs to be re-extracted for it.

> Unfortunately, there isn't a lot we can do about homebrew shipping a broken fluidsynth library.

Indeed, it does seem odd that homebrew ships a broken devel component for fluidsynth. I guess the runtime libraries work fine so no one noticed?

**Comment by kcgen** [comment]
@shermp , 

I looked at the order of these two commits: 
 1. Updating Staging's FluidSynth wrap to use 2.2.4-3 
 2. Setting the `fluidsynth:openmp` default option (which depends on FS 2.2.4-3)

And the order is correct; that is, we upgraded to 2.2.4-3 and only then subsequently used the new option:

### Dec 13 @ 8:13am - Staging's FluidSynth wrap updated to use 2.2.4-3 

![2022-02-24_14-23](https://user-images.githubusercontent.com/1557255/155617628-ee7d3dbb-98e3-4939-a9be-154c4c45e399.png)

### Dec 13 @ 9:41am - Using FS 2.2.4-3's `fluidsynth:openmp` option

![2022-02-24_14-22](https://user-images.githubusercontent.com/1557255/155617633-7cedfd22-4a0a-4c80-8935-15bf25d6f775.png)

So I believe the scenario should be impossible:

Scenario 1: User using an old Meson (<0.60) and using an old commit (prior to Dec 13) that doesn't include the FS 2.2.4-3 wrap or the use of the `fluidsynth:openmp` option:
- The old wrap in the repo doesn't support `fluidsynth:openmp`, but we don't use it.
- The old Meson isn't presented with conflicting options.
- Result: builds fine.

Scenario 2: User using an old Meson (<0.60) and an intermediate commit that _includes_ the FS 2.2.4-3 wrap but prior to the use of the `fluidsynth:openmp` option:
- The new wrap in the repo supports the `fluidsynth:openmp`, but we don't use it.
- The old Meson isn't presented with conflicting options.
- Result: builds fine.

Scenario 3: User using an old Meson (<0.60) and a later commit that _includes both_ the FS 2.2.4-3 wrap and the use of the `fluidsynth:openmp` option:
- The new wrap in the repo supports the `fluidsynth:openmp`, and we use it.
- The old Meson isn't presented with conflicting options because the wrap supports it.
- Result: builds fine.

Scenario 4: User using a new Meson (>=0.60) and using an old commit (prior to Dec 13) that doesn't include the FS 2.2.4-3 wrap or the use of the `fluidsynth:openmp` option:
- The old wrap in the repo doesn't support `fluidsynth:openmp`, but we don't use it.
- The new Meson isn't presented with conflicting options.
- Result: builds fine.

Scenario 5: User using a new Meson (>=0.60) and an intermediate commit that _includes_ the FS 2.2.4-3 wrap but prior to the use of the `fluidsynth:openmp` option:
- The new wrap in the repo supports the `fluidsynth:openmp`, but we don't use it.
- The new Meson isn't presented with conflicting options.
- Result: builds fine.

Scenario 6: User using a new Meson (>=0.60) and a later commit that _includes both_ the FS 2.2.4-3 wrap and the use of the `fluidsynth:openmp` option:
- The new wrap in the repo supports the `fluidsynth:openmp`, and we use it.
- The new Meson isn't presented with conflicting options because the wrap supports it.
- Result: builds fine.

So in terms of ordering things right in our repo, and commit can be used with any version of meson -- and our repo will checkout files that should build correctly.

So I believe the only way this happened in @PatTheMav's case is that the specific files within the `build/` area become inconsistent with the state of the repo's files. 

Last year, I added this wording to the build notes in a commit titled "Improve the build notes" (which made a bunch of other wording updates):

```
<snip> if you have unexpected build failures:

    ``` shell
    ccache -C
    git clean -fdx
    ```
```

Although this is pretty heavy handed, it would have also helped @PatTheMav blow away whatever inconsistent state had acrued in the non-repo derived/fetched files. 

But there were objections to this wording, so I removed it:

![2022-02-25_10-56](https://user-images.githubusercontent.com/1557255/155772745-423aa074-91d4-4440-864c-59f3e36b0e97.png)


It sounds like we need to re-add some kind of helper notes to help users who only periodically build get their environment back up to sync.  I'm not sure if that's `meson subprojects update --reset <wrap name>`, or if there's a stronger form.

Basically we want something like: _"If your build fails, then you might have stale cached/saved files, so please try refreshing it with command X, and then retry the build.  If that doesn't work, please open an issue"_

**Comment by eli-schwartz** [comment]
Scenario 7:

User builds prior to the FS 2.2.4-3 wrap. It works. The user `git pull`s, but the *old* subproject (a valid FS, just an old one) is already there, and gets picked up. 

Manually resetting the subproject via the command I mentioned will work around this.

**Comment by kcgen** [comment]
> User builds prior to the FS 2.2.4-3 wrap. It works. The user git pulls, but the old subproject (a valid FS, just an old one) is already there, and gets picked up.

After git pull'ing, the project's FluidSynth wrap will specify version 2.2.4-3. 

If Meson reads that wrap, it would see that there is no buit subproject of FluidSynth matching that version - and should fetch and build it per the wrap.

Obviously that's not happening though.

Is there a way we can tell Meson "Use the version specified in the wrap"?

**Comment by kcgen** [comment]
If Meson will just pick any old version of a subproject, then this doesn't match what I want, as a user.

I want Meson to fetch and build the requested version if it's not present (automatically, ideally) or fail with some help for the user (Your subproject X is version 1.0 but the repository wants 2.0, please get your subprojects back in sync by running `meson subprojects update --reset ...`)

**Comment by eli-schwartz** [comment]
As I explained before, the cache busting key Meson uses for this is the directory name the wrap is extracted to, which contains the FS version from the stable tag. It just doesn't contain the version of the meson.build patch overlay.

You'll always get the correct FS version, but if you update the meson.build while keeping the same FS version there is no obvious way to actually detect that that has happened.

I warmly encourage you to propose a method of cache busting for the meson.build patch overlay that doesn't involve force re-extracting it every single time you run meson "just in case".

**Comment by eli-schwartz** [comment]
> "Your Meson-managed subproject is version 1.0 but the wrap wants 2.0, so please update your subproject to use 2.0 using `meson subprojects update --reset` ..."

Let's rephrase that.

"Your Meson-managed subproject is version 2.0 and the wrap wants 2.0 also. You might think everything is good, and so does Meson, but the wrap also wants revision 3 of the meson.build patch, and you have revision 2. Please update your subproject using `meson subprojects update --reset`. No, we don't currently have a way to programmatically detect when to tell you this."

**Comment by kcgen** [comment]
> I warmly encourage you to propose a method of cache busting for the meson.build patch overlay that doesn't involve force re-extracting it every single time you run meson "just in case".

Indeed. We're in rare corner-case land where the user has an old build true, is knowledgable enough to building with Meson in the first place (but isn't aware of these more subtle issues) - and also isn't willing to clean or reclone fresh (I will admit, I've probably been tripped up by this a couple times across wrap updates - but I just cleaned and carried on).  Imposing any measurable cost (like re-extracting every build), is something I definitely don't want. 

>  Meson uses for this is the directory name the wrap is extracted to, which contains the FS version from the stable tag. It just doesn't contain the version of the meson.build patch overlay.

Maybe the answer is in there.

The subproject directory `subprojects/fluidsynth-2.2.4/` contains the upstream content per this:
``` ini
directory = fluidsynth-2.2.4
source_url = https://github.com/FluidSynth/fluidsynth/archive/v2.2.4.tar.gz
```

That's been modified per this:

``` ini
patch_filename = fluidsynth_2.2.4-3_patch.zip
patch_url = https://wrapdb.mesonbuild.com/v2/fluidsynth_2.2.4-3/get_patch
patch_hash = 2716afbb86046d817ab297b820f8f243d0a0779dbe4197ea4275f1e5b215ecce
```

So the directory `subprojects/fluidsynth-2.2.4/` isn't really upstream anymore; it's mangled by the patch. So my first approach would be to name the directory per the upstream directory plus abbreviated git-style patch revision (first 7 from patch_hash):

`subprojects/fluidsynth-2.2.4-patch-2716afb`

And then carry that directory name forward into whatever book keeping exists further into the build area (meson-info, meson-private, etc..) 

A second approach is a date compare. If the date of the source wrap file is newer than the extracted directory name, then we're almost guaranteed to be in this corner-case.

**Comment by eli-schwartz** [comment]
The directory name isn't created by meson, though. It comes from the upstream tagged tarball and is extracted without modification.

And FWIW:
> wraps that only update the patch revision can keep feeding off the 'clean' upstream directory without the need to redownload and extract the same upstream zip.

They shouldn't need to redownload since the upstream tarball is cached and checksummed so it should still be valid.

It would be the difference between re-extracting or doing a recursive copy. I am not sure there's a compelling reason based on time to choose one or the other, but copying means taking up twice as much space.

Checking the date might be promising. Although checking the meson.build file might be more reliable... except we don't know that that is the file being overlaid (it probably is...)

**Comment by kcgen** [comment]
> The directory name isn't created by meson, though. It comes from the upstream tagged tarball and is extracted without modification.

The extracted directory matches the `directory = ` specified in the .wrap (no?)

So if the .wrap includes a patch, the output from that operation becomes a new directory named `directory + "patch" + patch_abbreviation`, and then that's carried forward into the build area.

If that directory doesn't exist when building, then there's a version mismatch and the update --refesh routine can be rerun for the subpackage. 

> Although checking the meson.build file might be more reliable... except we don't know that that is the file being overlaid (it probably is...)

Nice refinement! No matter what, the subpackage's meson.build file (if it's overlaid or not), should always have an mtime equal or newer than the mtime of it's.wrap.  If the wrap is newer, then either the upstream zip has changed or the patch has been revised - and both cases would warrant an update.

**Comment by kcgen** [comment]
Let's get some actionable fixes to the documentation going as well. 
1. If the build fails with errors from the compiler (gcc/clang/msvc or linker, please open a new issue. 
2. If the build fails due to errors from Meson, please reconfirm all dependencies are installed. If the issue relates to a subpackage, please reset the affected subpackag with: `meson subprojects update --reset name-of-subpackage`, for example, `meson subprojects update --reset fluidsynth`
3. If that doesn't help, try resetting your build area with `git checkout -f main && git pull && git clean -fdx`. Worst case,  try a fresh clone of the repo. 

@PatTheMav , will do you've suggested and knock out FluidSynth from the list of macOS dependencies (until Brew fixes that up). The wrap provides an optimized and lean alternative, so is a perfectly good fallback.
